Add MenuManager and integrate it into VGMFileListView

mikelow opened the pull request:
As [promised](https://github.com/vgmtrans/vgmtrans/pull/417#issuecomment-1858931378), this PR adds batch conversion so that when multiple VGMFiles are selected, they can all be saved in a single command. But wait, there's more! I didn't just tack this on to VGMFileListView. I wrote a generalized solution that will hopefully make it easy to create menus for any type across the entire app.

## The core classes: <br>MenuManager, Command, CommandContext, and CommandContextFactory

This PR introduces a new "service" for menu generation called `MenuManager`. The idea is that you can register a set of "Commands" to any type you so desire, where a `Command` represents a menu action, and then pass an instance of any type to `MenuManager::GetCommands()` to retrieve the eligible `Command`s of the object. The Command object, in turn, can execute the action.

Commands require data, of course. For example, the command "Save as MIDI" needs to know what file(s) to save and what path to save them to. Hence, we need to know not only what Commands an object can execute but also the data required by them and a decoupled mechanism for passing data to them. Enter the `CommandContext` class, which provides data to the `Command`.

To actually run a command, the `Command` class provides the `Execute()` method which requires a `CommandContext` parameter. A `Command` subclass knows to expect a specific `CommandContext` subclass, and will dynamic_cast the parameter to the expected type. So to use a command, we need to create the appropriate CommandContext and populate it with data. To do so, the Command provides the `GetContextFactory()` method, which returns a subclassed `CommandContextFactory` instance. 

A `CommandContextFactory` provides both the information on the data required by a `CommandContext` and the means to create it. The method `CreateContext()` takes a `PropertyMap` - a key/value map - and returns the specialized CommandContext. The `GetPropertySpecifications()` method returns a list of `PropertySpecification`s that detail the keys and corresponding value types expected to be entered into the PropertyMap passed to `CreateContext()`.

Ok, so the process is:

1. At app startup, register Commands for types by calling `MenuManager::RegisterCommands()`
2. Get the Commands for an object by calling `MenuManager::GetCommands()`
3. To use a Command, first generate the CommandContextFactory by calling `Command::GetContextFactory()`
4. Get the specifications of required properties by calling `CommandContextFactory::GetPropertySpecifications()`
5. Populate a `PropertyMap` with the required properties.
6. Create a `CommandContext` by calling `CommandContextFactory::CreateContext()` with the populated `PropertyMap`
7. Execute the command by calling `Command::Execute()` with the `CommandContext`.

## CreateMenuForItems() and FindCommonCommands()

"Wait a second," you say, "I've been hoodwinked. Why, for all this devastatingly technical jibber jabber, you still haven't addressed the promised batch conversion feature." Now you just slow down there, and quit with those damned accusations. This PR does indeed address batch conversions. Why I wrote a couple functions for that very purpose.

There's the `MenuManager::FindCommonCommands()` function, which takes a vector of instance pointers and returns a vector of commands common to all of the instances. For example, if I pass a list of three VGMInstrSet instances, it will return commands including conversion to DLS and SF2, because they all offer that functionality. If I pass a VGMInstrSet and a VGMSeq, it will only give me the commands to close the files or save them in their original format.

I've also written the `MenuManager::CreateMenuForItems()`, which uses `FindCommonCommands()` to generate a QMenu. It also accepts a vector a instance pointers.

## Handling Polymorphism in Command Registration

One of the difficulties I ran into is that C++ offers limited RTTI (run-time type information). Specifically, while we can get the type name of an instance using typeid(), we cannot get its class hierarchy. In our use case, it's important that we be able to register commands to a common parent class, like VGMSeq, VGMInstrSet, and VGMSampColl. But if `typeid()` will only tell us that a VGMFile instance is a `RareSnesSeq` and not a `VGMSeq`, what are we to do?

My solution is to have two registries within MenuManager. The first is a simple map of type_index to a vector of Command which we will always check first. The second is a vector of pairs, where the first value is a function that tests whether an instance can be dynamic_cast to a type, and the second is the vector of Commands for that type. This is obviously slower than the map lookup, since we need to execute every test function until we get a match, but it works. The other downside is that we need to limit this functionality to a specific base class, as we can't use `dynamic_cast` on a void*; it must know the base class. For our needs, this functionality works on VGMItem*. Of course, we could expand this to more base types if needed.

## Services and Dependency Injection

You'll notice that I called MenuManager a "service" and put it in a new "services" directory. The idea here is to move away from using singletons - especially the VGMRoot singleton - and toward dependency injection. I'm not actually using dependency injection in this PR; the MenuManager is owned by VGMFileList because I didn't want to increase the scope of this PR. Howeve,r the way this is structured allows us to use MenuManager as an injected dependency. I'm calling these injected dependencies "services" but I'm open to using another term.

## Is this an example of overengineering?

I wrestled with this concern. I don't think so. A major problem with the app is that most functionality is hidden behind right-click context menus. This is poor design. Users will be confused and/or never discover functionality. We should be updating the application menu with the same contextual menu items as a right click. I think this will make it easier to do that across the app.

## How Has This Been Tested?
Tested on MacOS Sonoma x64/arm64 and Windows 10.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.

sykhro commented:
Great work, it's nice to see things moving again!
I will review this in more detail tomorrow but it looks good at first glance.

> One of the difficulties I ran into is that C++ offers limited RTTI (run-time type information). Specifically, while we can get the type name of an instance using typeid(), we cannot get its class hierarchy. In our use case, it's important that we be able to register commands to a common parent class, like VGMSeq, VGMInstrSet, and VGMSampColl. But if typeid() will only tell us that a VGMFile instance is a RareSnesSeq and not a VGMSeq, what are we to do?

With the current architecture of things I can't really think of simpler solutions. A clean-ish C++ way of doing this is through type-erasure (something like [this](https://gist.github.com/mrkline/114fa9963a6615231c40247e129c21b4)).
It would essentially be the same thing behind the scenes, it just makes the "figuring out the type" thing less explicit. 

>  The idea here is to move away from using singletons - especially the VGMRoot singleton - and toward dependency injection. ... I'm calling it a service, but I'm happy to call it something else.

This is the direction I wanted to take too (e.g. logging in the `refactor` branch, even if it's actually a singleton in disguise...). `services` sounds fine

mikelow commented:
> With the current architecture of things I can't really think of simpler solutions. A clean-ish C++ way of doing this is through type-erasure (something like [this](https://gist.github.com/mrkline/114fa9963a6615231c40247e129c21b4)).
> It would essentially be the same thing behind the scenes, it just makes the "figuring out the type" thing less explicit.

I'm still trying to grok this. I found the author's talk [Inheritance is the Base Class of Evil](https://www.youtube.com/watch?v=bIhUE5uUFOA). Think I'm going to need a good sleep before I can fully follow it.

If you know how to apply this pattern to improve the implementation, please let me know! Sounds like it's a powerful way to use polymorphism without class inheritance. My feeble brain hasn't grasped yet how to put it into practice.

> This is the direction I wanted to take too (e.g. logging in the refactor branch, even if it's actually a singleton in disguise...). services sounds fine

Excellent. Not surprised we're on the same page. If you know of a more common naming convention, let me know. If we go with "service", I'm debating renaming `MenuManager` to `MenuService`.

sykhro commented:
> I'm still trying to grok this. I found the author's talk [Inheritance is the Base Class of Evil](https://www.youtube.com/watch?v=bIhUE5uUFOA). Think I'm going to need a good sleep before I can fully follow it.

Hehe, yes, it's not straightforward; I hope I didn't send you down a rabbit hole! If you're open to considering c++-idiom alternatives, there's [double dispatch](https://eli.thegreenplace.net/2016/a-polyglots-guide-to-multiple-dispatch/) too (arguably less complex than the above)

> If you know how to apply this pattern to improve the implementation, please let me know! 

The problem is finding the common interface for the class to hide. If we have a `CommandTarget` as `object_t`, then the common function could be `CommandTarget::execute`, calling a free function e.g. `executeCommand`. So you could have `void executeCommand(VGMSeq *, ...` which should be picked automatically when you invoke `executeCommand` on a  `RareSnesSeq`. But then it becomes more complicated to.. actually run the command. Welp. I will think about this some more

> Excellent. Not surprised we're on the same page. If you know of a more common naming convention, let me know. If we go with "service", I'm debating renaming `MenuManager` to `MenuService`.

I honestly don't know. I have seen *Manager and *Service in codebases, they are the go-to "I don't know how to name this thing but it feels important". I have to say that *Service sounds like something that's running in the background, but I don't think we'll ever have anything of the sort.

mikelow commented:
One of my goals with the MenuManager class is to 

> Hehe, yes, it's not straightforward; I hope I didn't send you down a rabbit hole! If you're open to considering c++-idiom alternatives, there's [double dispatch](https://eli.thegreenplace.net/2016/a-polyglots-guide-to-multiple-dispatch/) too (arguably less complex than the above)

That is a cool pattern, but it's not clear to me how I would utilize it. If there were a way, it seems it would requiring updating the classes which register with the MenuManager, something I'm trying to avoid.

> The problem is finding the common interface for the class to hide. If we have a CommandTarget as object_t, then the common function could be CommandTarget::execute, calling a free function e.g. executeCommand. So you could have void executeCommand(VGMSeq *, ... which should be picked automatically when you invoke executeCommand on a RareSnesSeq. But then it becomes more complicated to.. actually run the command. Welp. I will think about this some more

I'll go over the material and try to make sense of all of this. The double dispatch pattern was easy to understand and explained well. The type-erasure stuff, not so much.

> I honestly don't know. I have seen *Manager and *Service in codebases, they are the go-to "I don't know how to name this thing but it feels important". I have to say that *Service sounds like something that's running in the background, but I don't think we'll ever have anything of the sort.

I've seen "service" used for dependency injected modules outside of C++. I suppose there isn't any standard. My qualm with "manager" is that it's too specific in its meaning. Granted, service does have that operating system "background service" association.

mikelow commented:
> The problem is finding the common interface for the class to hide. If we have a CommandTarget as object_t, then the common function could be CommandTarget::execute, calling a free function e.g. executeCommand. So you could have void executeCommand(VGMSeq *, ... which should be picked automatically when you invoke executeCommand on a RareSnesSeq. But then it becomes more complicated to.. actually run the command. Welp. I will think about this some more

Well the way I was thinking about it, the common interface would be a getCommands() function that returns a different list of Commands for each type. The problem I'd like to solve is the coupling of MenuManager with VGMItem; so that we can register commands for other parent classes and still have them work for their subclasses. Even so I can't wrap my head around this design pattern - I've been fiddling around, but the example code is too different from what I'm trying to do and I find it unintuitive.

sykhro commented:
> it would requiring updating the classes that register with the MenuManager, which is something I'm trying to avoid.

ah ok, fair enough
 
 > the common interface would be a getCommands() function that returns a different list of Commands for each type
 
 Juggling types like this in c++ is basically only possible as long as the information you need is known at compile time
 
> The problem I'd like to solve is the coupling of MenuManager with VGMItem; so that we can register commands for other parent classes and still have them work for their subclasses. Even so I can't wrap my head around this design pattern - I've been experimenting, but the example code is too different from what I'm trying to do and I find it unintuitive.

I have tried but didn't end up with anything useful. I'm finding it hard to find something else that would work with everything that we have in place right now, so this is ok for me

mikelow commented:
> it would requiring updating the classes that register with the MenuManager, which is something I'm trying to avoid.

Well, I'm not _completely_ opposed to this.  I'm trying to retain the separation between UI logic and the core logic. One easy way to solve this is to just make a dummy base class everything inherits from. Then dynamic_cast works with all of them. Doesn't seem like a great solution, but neither is hardcoding VGMItem.

I didn't look into whether we could register the parent class somehow instead of hardcoding it. That would be cleaner.

> Juggling types like this in c++ is basically only possible as long as the information you need is known at compile time

Yeah, it occurred to me the type-erasure design probably can't fix this problem. At that point, I think it would just get us constant time resolution of the polymorphic class, as opposed to O(n) where we iterate over a vector of functions. I'm all for it, if we can figure out a way, but it doesn't solve what I consider the bigger design issue.

Practically, I don't foresee the O(n) lookup being a hindrance, as all leaf subclasses are still resolved in constant time, and I don't expect there will be many more types that need to be resolved via the checkFunc vector. Also, C++ is fast.

> I have tried but didn't end up with anything useful. I'm finding it hard to find something else that would work with everything that we have in place right now, so this is ok for me

👍 Don't hold back your critiques.

mikelow commented:
> I didn't look into whether we could register the parent class somehow instead of hardcoding it. That would be cleaner.

I've just added a commit that does exactly this. Now RegisterCommands() takes an optional second template argument to specify the base class for which to register the commands (I've removed the singular RemoveCommand() method). The `commandsForCheckedType` property is now a map, associating a base class type to its former type - a vector of pairs, where the first item is a function to test if an instance belongs to class, and the second item is a vector of Commands for that class.

The only difference for client usage now is that CreateMenuForItems() expects a template argument to specify which base type to use for dynamic type lookup. This type will need to match whichever base type the commands were registered with. Of course, if the type was not registered with a base type, it will not matter.

I've also added some meta programming magic in the form of `has_getname`, which, when used with constexpr, let's us check if a template argument has the method GetName() at compile time. Now, `CreateMenuForItems()` will work for any type, whereas before it would have failed if the type did not define GetName(). If GetName() is missing, it will simply _not_ provide a suggested save file name when opening a save file dialog.

mikelow commented:
Waiting for approval. Are you still interested in exploring refactoring?

sykhro commented:
I was thinking about this earlier. imo this solution is okay, I feel like it's better to get the ball rolling rather than lacking features forever :)

sykhro commented on src/ui/qt/services/commands/SaveCommands.h:
```diff
@@ -0,0 +1,206 @@
+    if (alwaysSaveToDir) {
+      // alwaysSaveToDir indicates we were given a directory path and Save() also expects a directory path
+      for (auto file : files) {
+        auto specificFile = dynamic_cast<TSavable*>(file);
```
I only have a nitpick: `Savable` or `Saveable`?

mikelow replied:
Interesting lexical question. I had to research this. It seems both "savable" and "saveable" are acceptable spellings. I've seen some sources claim the meaning of the two words is different, but best I can tell that's not true. Savable seems to be more common in American English. The American English MacOS spell checker doesn't even like "saveable".

There are, however, twice as many usages of "saveable" in github code than "savable".  If "Savable" rubs you the wrong way, I will happily change it to "Saveable". 

<img width="306" alt="image" src="https://github.com/vgmtrans/vgmtrans/assets/1659535/9a1b78a0-ef84-4dae-a08a-18043ab66b6a">

sykhro replied:
ah, interesting, it felt like a typo to my non-native brain. let's leave it like this then

sykhro commented on src/ui/qt/workarea/VGMFileListView.cpp:
```diff
@@ -26,7 +26,7 @@ VGMFileListModel::VGMFileListModel(QObject *parent) : QAbstractTableModel(parent
 QVariant VGMFileListModel::data(const QModelIndex &index, int role) const {
   if (!index.isValid()) {
-    return QVariant();
+    return {};
```
This was `QVariant()` because codegen with `{}` used to be horrendous (something about copy construction in Qt, I don't remember). Did you observe a performance difference?

mikelow replied:
Not sure if I noticed a performance difference, but I'll switch it back if that's even potentially an issue. To be clear, I made this change because of a clang-tidy warning, which I assumed meant it would be functionally equivalent.

_"Clang-tidy: Void repeating the return type from the declaration; use a braced initializer list instead"_

sykhro replied:
I had a look at QVariant and it looks saner nowadays, looks like this is no longer an issue

sykhro commented on src/ui/qt/workarea/VGMFileListView.cpp:
```diff
@@ -154,48 +154,24 @@ void VGMFileListView::onHeaderSectionResized(int index, int oldSize, int newSize
 
-  VGMFile *file = qtVGMRoot.vVGMFile[element.row()];
-  if (!file) {
+  auto selectedFiles = make_shared<vector<VGMFile*>>();
```
Any reason why this can't be a `unique_ptr` (or better, just a `vector` value?)

mikelow replied:
Because that would make too much sense. I need to needlessly complicate this.

😉 I'll change it.

mikelow replied:
Oops, I forgot that I did have a reason for making this a shared_ptr: CreateMenuForItems() is passing a closure to QMenu::addAction() which utilizes the vector of VGMFile*. This closure will be invoked beyond the lifetime of selectFiles` scope.

If you think it would be cleaner to copy the vector, though. I could do that.

sykhro replied:
Ah, whoops, makes sense that it would go past the selectFiles lifetime. Then I'd say that a shared pointer is the proper abstraction for ownership here

sykhro commented on src/ui/qt/services/MenuManager.h:
```diff
@@ -0,0 +1,302 @@
+ * @return A QMenu populated with actions for the common commands, in the same order as they were registered for the first item
+   */
+  template <typename Base, typename T = Base>
+  QMenu* CreateMenuForItems(shared_ptr<vector<T*>> items) {
```
since the vector isn't really owned by this,  `std::vector<T*> &` is more appropriate than `shared_ptr`

mikelow replied:
💯

mikelow replied:
See above.

mikelow commented:
> I was thinking about this earlier. imo this solution is okay, I feel like it's better to get the ball rolling rather than lacking features forever :)

If you think this code smells, I don't mind taking it back to the drawing board. I just don't know how it could be refactored.

sykhro commented:
> > I was thinking about this earlier. imo this solution is okay, I feel like it's better to get the ball rolling rather than lacking features forever :)
> 
> If you think this code smells, I don't mind taking it back to the drawing board. I just don't know how it could be refactored.



> > I was thinking about this earlier. imo this solution is okay, I feel like it's better to get the ball rolling rather than lacking features forever :)
> 
> If you think this code smells, I don't mind taking it back to the drawing board. I just don't know how it could be refactored.

It took me a bit to understand it but I wouldn't say it's bad. We are in a very dynamic situation so any other solution would be very similar imo. I'm happy to merge this
